text/0724-road-to-typescript.md

```diff
@@ -0,0 +1,236 @@
+
+
+#### Recommendations
+
+Since the authoring and support policies we adopt for Ember itself will likely be adopted in an _ad hoc_ way, it may be useful to provide official *recommendations* for authors. These might include:
+
+- Semantic Versioning guidelines
+- tooling recommendations (in line with the defaults for the build pipeline)
+- documentation guidelines
+
+The existence of these guidelines is not a requirement for Ember's official adoption of TypeScript, but the adoption process would be a prime opportunity to generate these artifacts.
+
+
+### Semantic Versioning and Supported TypeScript Versions
+
+Ember and TypeScript have fundamentally different views on Semantic Versioning (SemVer).
+
+Ember has a deep commitment to minimizing breaking changes in general, and to strictly following SemVer when breaking changes *are* made. (The use of lockstep versioning for Ember Data and Ember CLI complicates this commitment to a substantial degree, but that complication is outside the scope of this RFC.)
+
+TypeScript explicitly *does not* follow SemVer. TypeScript's core team argues that *every* change to the compiler is a breaking change, and that SemVer is therefore meaningless. (We do not agree with this characterization, but are also uninterested in arguing it. This RFC takes the TypeScript team's position as a given.) Accordingly, every TypeScript point release may be a breaking change, and "major" numbers for releases signify nothing beyond having reached `x.9` in the previous cycle.
+
+For TypeScript to be a first-class citizen of the Ember ecosystem, we need:
+
+- a policy defining what constitutes a breaking change for consumers of a library which publishes types, including Ember’s core libraries
+- tooling to detect breaking changes in types—whether from refactors, or from new TypeScript releases—and to minimize the amount of churn from breaking changes in TypeScript
+- a general and widely-adopted policy for supported TypeScript versions
+
+Once all three of those elements are adopted, end users will be able to have equally high confidence in the stability of published types as they do in their runtime code.
+
+
+## How we teach this
+
+This RFC intentionally defers most of the analysis of teaching work to an RFC dedicated to the question. At a high level, once the roadmap is complete, the Ember blog should announce support, and Ember’s docs should explicitly state that TypeScript is officially supported and show TypeScript examples alongside JavaScript examples.
+
+It is critical that these materials (including the blog post) emphasize that TypeScript support is *additive*, not a replacement for JavaScript, as discussed above in [Non-Goals](#non-goals).
+
+
+## Drawbacks
+
+Adding TypeScript support imposes an additional maintenance burden on all contributors to Ember:
+
+- documentation must be kept up to date in both JavaScript and TypeScript
```

Comment:
as part of the build step of deploying documentation, docs could be written in TS, and then, at build time, have types tripped away? 
I'm imagining like how Microsoft has their C# / F# / VB.net docs, where there are tabs to switch between examples.
For us, since TS is a superset of JS, authoring in TS means we run each code sample through babel at build time to strip the types, and then output a JS "tab" as well?

(of course now I can't find any examples of this, but I promise they exist somewhere!)

Reply:
Plus 1 on deriving JS docs from TypeScript.

Reply:
This is a great callout, and very much along the lines of what I and others in the strike team have considered. It is, however, also the kind of thing we should discuss *on a dedicated docs RFC*, as this RFC intentionally does not address it, per the introduction! Marking this resolved accordingly, but if you identify gaps around what we should or shouldn't say about docs as part of this "roadmap" RFC, please do open further issues!